panic when load entries in WAL

I start a 3-node cluster. Do some requests. Restart all of them. Restart all of them. When I start some member:

```
2014/11/26 06:01:38 no data-dir provided, using default data-dir ./infra1.etcd
2014/11/26 06:01:38 etcd: listening for peers on http://172.31.39.94:7001
2014/11/26 06:01:38 etcd: listening for client requests on http://172.31.39.94:4001
2014/11/26 06:01:38 etcdserver: recovering from snapshot at index 15334753
2014/11/26 06:01:38 etcdserver: name = infra1
2014/11/26 06:01:38 etcdserver: data dir = infra1.etcd
2014/11/26 06:01:38 etcdserver: snapshot count = 10000
2014/11/26 06:01:38 etcdserver: advertise client URLs = http://172.31.39.94:4001
2014/11/26 06:01:38 etcdserver: loaded peers from snapshot: infra1=http://172.31.39.94:7001,infra2=http://172.31.39.92:7002,infra3=http://172.31.39.93:7003
panic: runtime error: slice bounds out of range

goroutine 16 [running]:
runtime.panic(0x7a6fa0, 0xa08faf)
        /usr/local/go/src/pkg/runtime/panic.c:279 +0xf5
github.com/coreos/etcd/wal.(*WAL).ReadAll(0xc208004360, 0xc2080425c0, 0x14, 0x20, 0xf, 0xddcc6eb479e9a80b, 0xe9fd61, 0x0, 0x0, 0x0, ...)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/wal/wal.go:177 +0x79a
github.com/coreos/etcd/etcdserver.readWAL(0xc2080ca6a0, 0xf, 0xe9fd61, 0xc208004360, 0x44f1dc, 0xc2080181e0, 0x0, 0x0, 0x0, 0x0, ...)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:933 +0x1b0
github.com/coreos/etcd/etcdserver.restartNode(0xc20804a510, 0xe9fd61, 0xc2080042a0, 0x1, 0x0, 0x0, 0xc208018791)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:919 +0x87
github.com/coreos/etcd/etcdserver.NewServer(0xc20804a510, 0x3, 0x0, 0x0)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:287 +0x1560
github.com/coreos/etcd/etcdmain.startEtcd(0x0, 0x0, 0x0)
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:291 +0x1572
github.com/coreos/etcd/etcdmain.Main()
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:176 +0x38b
main.main()
        /home/ubuntu/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/main.go:33 +0x1a

goroutine 19 [finalizer wait]:
runtime.park(0x412e00, 0xa0cfb0, 0xa0b249)
        /usr/local/go/src/pkg/runtime/proc.c:1369 +0x89
runtime.parkunlock(0xa0cfb0, 0xa0b249)
        /usr/local/go/src/pkg/runtime/proc.c:1385 +0x3b
runfinq()
        /usr/local/go/src/pkg/runtime/mgc0.c:2644 +0xcf
runtime.goexit()
        /usr/local/go/src/pkg/runtime/proc.c:1445
```

the last snapshot is 0000000000000005-0000000000e9fd61.snap, the last wal is 00000000000005ee-0000000000e9fd57.wal

some debug info:

```
open at 15334753 from infra1.etcd/wal/00000000000005ee-0000000000e9fd57.wal
read entry with index 15334729
read entry with index 15334730
read entry with index 15334731
read entry with index 15334732
read entry with index 15334733
read entry with index 15334734
read entry with index 15334735
read entry with index 15334736
read entry with index 15334737
read entry with index 15334738
read entry with index 15334739
read entry with index 15334740
read entry with index 15334741
read entry with index 15334742
read entry with index 15334754
```

--- a/raft/log.go
+++ b/raft/log.go
@@ -216,8 +216,8 @@ func (l *raftLog) snap(d []byte, index, term uint64, nodes []uint64) {
 }
 
 func (l *raftLog) restore(s pb.Snapshot) {
-	l.ents = []pb.Entry{{Term: s.Term}}
-	l.unstable = s.Index + 1
+	l.ents = []pb.Entry{{Index: s.Index, Term: s.Term}}
+	l.unstable = s.Index
 	l.committed = s.Index
 	l.offset = s.Index
 	l.snapshot = s

--- a/raft/log_test.go
+++ b/raft/log_test.go
@@ -501,8 +501,8 @@ func TestLogRestore(t *testing.T) {
 	if raftLog.committed != index {
 		t.Errorf("comitted = %d, want %d", raftLog.committed, index)
 	}
-	if raftLog.unstable != index+1 {
-		t.Errorf("unstable = %d, want %d", raftLog.unstable, index+1)
+	if raftLog.unstable != index {
+		t.Errorf("unstable = %d, want %d", raftLog.unstable, index)
 	}
 	if raftLog.term(index) != term {
 		t.Errorf("term = %d, want %d", raftLog.term(index), term)

